### PR TITLE
Added user permissions

### PR DESCRIPTION
**Pull request by Cheshiriks**
Added user permissions

### CODE REVIEW

**Review by sanyavertolet — APPROVED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * @property isPermittedCreateContest permission for create contests
+ * @property isPermittedToBulkUpload permission for upload COSV files
+ * @property organizationName name of organization
+ */
+@Serializable
+data class UserPermissions(
+    val isPermittedCreateContest: Boolean = false,
+    val isPermittedToBulkUpload: Boolean = false,
```
Comment (nulls) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:13]: user allowed to upload COSV and create contest only in organization. Organization should be not null

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import kotlinx.serialization.Serializable
+
+/**
+ * @property isPermittedCreateContest permission for create contests
+ * @property isPermittedToBulkUpload permission for upload COSV files
+ * @property organizationName name of organization
+ */
+@Serializable
+data class UserPermissions(
```
Comment (nulls) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:11]:
```suggestion
data class UserPermissionsInOrganization(
```

Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:11]: I'm thinking of making this a more general class that won't just be tied to organizations.

Comment (akuleshov7) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:11]: my suggestion is to keep these flags on organization level, why not, @nulls 
?

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    val isPermittedCreateContest: Boolean = false,
+    val isPermittedToBulkUpload: Boolean = false,
+    val organizationName: String? = null,
```
Comment (akuleshov7) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:14]: remove this field and move it to flags (make it a complex)

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt (hunk 2 above)
Comment (nulls) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:11]: In future, those flags can be different for different users, but now they are common for all users in one organization.
I think to have them in separate object will simplify this change in future

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt (hunk 1 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt]: Changed it

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt]: Fixed

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:10]: Changed it

Comment (nulls) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:10]: yes, but 
> user allowed to upload COSV and create contest only in organization

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import kotlinx.serialization.Serializable
+
+/**
+ * @property isPermittedCreateContest permission for create contests in organizations
+ * @property isPermittedToBulkUpload permission for upload COSV files in organizations
+ */
+@Serializable
+data class UserPermissions(
+    val isPermittedCreateContest: Map<String, Boolean> = emptyMap(),
```
Comment (nulls) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:11]: suggestion to split it to two classes:
1. ```kotlin
    @Serializable
     data class UserPermissions(
         canCreateContest: Boolean,
         canDoBulkUpload: Boolean,
     )
    ```
2. ```kotlin
    data class AllUserPermissions(
        val inOrganization: Map<String, UserPermissions>,
    )
    ```

It allows to extend UserPermissions in single place using only Organization and allows to add another permissions per project\vulnerability\contest and etc

Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt]: Ok, agree. Fixed it

**Review by nulls — APPROVED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.saveourtool.save.info
+
+import kotlinx.serialization.Serializable
+
+/**
+ * @property inInOrganizations user permissions in organizations
+ */
+@Serializable
+data class UserPermissions(
+    val inInOrganizations: Map<String, UserPermissionsInOrganization> = emptyMap(),
```
Comment (nulls) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt:10]:
```suggestion
    val inOrganizations: Map<String, UserPermissionsInOrganization> = emptyMap(),
```

**Review by nulls — APPROVED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserPermissions.kt]: Fixed

**Pull request merged**